// @serverless-stack/resources/dist/util/apiGatewayV2Cors.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCorsConfig = void 0;
const apig = __importStar(require("@aws-cdk/aws-apigatewayv2-alpha"));
const duration_1 = require("./duration");
function buildCorsConfig(cors) {
    // Handle cors: false
    if (cors === false) {
        return;
    }
    // Handle cors: true | undefined
    else if (cors === undefined || cors === true) {
        return {
            allowHeaders: ["*"],
            allowMethods: [apig.CorsHttpMethod.ANY],
            allowOrigins: ["*"],
        };
    }
    // Handle cors: apig.CorsPreflightOptions
    else {
        return {
            allowCredentials: cors.allowCredentials,
            allowHeaders: cors.allowHeaders,
            allowMethods: (cors.allowMethods || []).map((method) => apig.CorsHttpMethod[method]),
            allowOrigins: cors.allowOrigins,
            exposeHeaders: cors.exposeHeaders,
            maxAge: cors.maxAge && (0, duration_1.toCdkDuration)(cors.maxAge),
        };
        cors;
    }
}
exports.buildCorsConfig = buildCorsConfig;
