// @serverless-stack/resources/dist/BaseSite.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuildCmdEnvironment = exports.buildErrorResponsesFor404ErrorPage = exports.buildErrorResponsesForRedirectToIndex = void 0;
const cdk = __importStar(require("aws-cdk-lib"));
function buildErrorResponsesForRedirectToIndex(indexPage) {
    return [
        {
            httpStatus: 403,
            responsePagePath: `/${indexPage}`,
            responseHttpStatus: 200,
        },
        {
            httpStatus: 404,
            responsePagePath: `/${indexPage}`,
            responseHttpStatus: 200,
        },
    ];
}
exports.buildErrorResponsesForRedirectToIndex = buildErrorResponsesForRedirectToIndex;
function buildErrorResponsesFor404ErrorPage(errorPage) {
    return [
        {
            httpStatus: 403,
            responsePagePath: `/${errorPage}`,
        },
        {
            httpStatus: 404,
            responsePagePath: `/${errorPage}`,
        },
    ];
}
exports.buildErrorResponsesFor404ErrorPage = buildErrorResponsesFor404ErrorPage;
/////////////////////
// Helper Functions
/////////////////////
function getBuildCmdEnvironment(siteEnvironment) {
    // Generate environment placeholders to be replaced
    // ie. environment => { API_URL: api.url }
    //     environment => API_URL="{{ API_URL }}"
    //
    const buildCmdEnvironment = {};
    Object.entries(siteEnvironment || {}).forEach(([key, value]) => {
        buildCmdEnvironment[key] = cdk.Token.isUnresolved(value)
            ? `{{ ${key} }}`
            : value;
    });
    return buildCmdEnvironment;
}
exports.getBuildCmdEnvironment = getBuildCmdEnvironment;
