// @serverless-stack/resources/dist/GraphQLApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLApi = void 0;
const Api_1 = require("./Api");
const _1 = require(".");
class GraphQLApi extends Api_1.Api {
    constructor(scope, id, props) {
        // Validate server
        if (!props.server) {
            throw new Error(`Missing "server" in the "${id}" GraphQLApi`);
        }
        if ("routes" in props) {
            throw new Error(`Please use the "server" option instead of the "routes" to configure the handler for the "${id}" GraphQLApi`);
        }
        /*
        if (props.codegen) {
          const app = App.of(scope) as App;
          if (!app.local) {
            const result = spawn.sync(
              "npx",
              ["graphql-codegen", "-c", props.codegen],
              {
                stdio: "inherit",
              }
            );
            if (result.status !== 0) {
              throw new Error(
                `Failed to generate the schema for the "${id}" GraphQLApi`
              );
            }
          }
        }
        */
        const rootPath = props.rootPath || "/";
        super(scope, id, Object.assign(Object.assign({}, props), { defaultPayloadFormatVersion: props.defaultPayloadFormatVersion || _1.ApiPayloadFormatVersion.V1, routes: {
                [`GET ${rootPath}`]: props.server,
                [`POST ${rootPath}`]: props.server,
            } }));
        this.rootPath = rootPath;
        this.codegen = props.codegen;
    }
    get serverFunction() {
        const serverFn = this.getFunction(`GET ${this.rootPath}`);
        // This should never happen
        if (!serverFn) {
            throw new Error(`Failed to get "serverFunction" in the "${this.node.id}" GraphQLApi`);
        }
        return serverFn;
    }
    // Note: We want to create 1 Lambda handling both the GET and POST request.
    //       This design is based on this discussion on GitHub
    //       https://github.com/serverless-stack/serverless-stack/issues/601
    // Also Note: We cannot use the "ANY /" route because if authorization
    //            were provided, the OPTIONS route will be protected. This
    //            causes CORS to fail.
    // Solution: We will override the createFunctionIntegration() function, and
    //           it will re-use the same Route Integration for all routes.
    createFunctionIntegration(scope, routeKey, routeProps, postfixName) {
        if (!this.lambdaIntegration) {
            this.lambdaIntegration = super.createFunctionIntegration(scope, routeKey, routeProps, postfixName);
        }
        return this.lambdaIntegration;
    }
    getConstructMetadata() {
        const parent = super.getConstructMetadata();
        return Object.assign(Object.assign({}, parent), { data: Object.assign(Object.assign({}, parent.data), { graphql: true, codegen: this.codegen }) });
    }
}
exports.GraphQLApi = GraphQLApi;
